### ISSUE TITLE
add config settings for package versions (e.g. titiler-pgstac==0.5.1)

### ISSUE THREAD

**Issue by hrodmn** [developmentseed/eoapi-cdk#85]
Right now the specific versions of the critical python packages are hard-coded in the runtime files

e.g.
https://github.com/developmentseed/eoapi-cdk/blob/1e39ddb54597754d4d621313e7ce34eebe9a8a6f/lib/titiler-pgstac-api/runtime/requirements.txt#L1

I have been following along over in https://github.com/developmentseed/eoAPI/pull/144 and see pattern you follow for overriding the runtime. I suppose I could use this method to achieve my goal but it seems heavy-handed for my goal of staying up-to-date on dependency upgrades!

https://github.com/developmentseed/eoAPI/pull/144/files#diff-5e565e714fe20093460499e9958e38735b99628ac37b603c30e2c8af61aee726R169-R184

One of the things that is very convenient about eoapi-cdk is that it makes it possible to just import the cdk modules without maintaining separate runtime scripts, so adding some configurability for the deployment of the packages would be nice!

Would it make sense to be able to set the versions of packages like titiler-pgstac and tipg as settings in the config? Maybe it would be tricky since some of the handler code might be dependent on version-specific details, but it seems like a way to decouple versions of eoapi-cdk from versions of the family of dependencies.

**Comment by emileten** [comment]
I think it it would be a useful option, and you do make me realize that the current situation is undesirable in that it forces a user to either use the versions that are hard-coded here, or do a lot of work to inject 'custom' runtimes. 

I did encounter this once in the past : 

> Maybe it would be tricky since some of the handler code might be dependent on version-specific details

Where updating the titiler version broke its handler, but I think since then we improved the way we wrote it, and in principle it should be as thin as possible and allow for the kind of feature you describe.

**Comment by emileten** [comment]
@hrodmn I just saw your PR. I have a quick question. Sorry that I had not thought of asking that question when writing my first reaction above. 

We have to update the package versions in the default deployment anyways, regardless of the feature we describe here, because some defaults are getting outdated. I just filed an issue for this, I was hoping to get to it soon : https://github.com/developmentseed/eoapi-cdk/issues/88. 

If these were updated, and updated frequently in the future, would your issue be solved, or would the feature we described above still be helpful ?

**Comment by hrodmn** [comment]
> If these were updated, and updated frequently in the future, would your issue be solved, or would the feature we described above still be helpful ?

@emileten I think it would still be useful because it would be good to be able to control the version of packages like `titiler-pgstac` which occasionally introduce breaking changes to the API. e.g. parameter name changes like `coord-crs` -> `coord_crs` in [v0.7.0](https://github.com/stac-utils/titiler-pgstac/blob/main/CHANGES.md#070-2023-09-28). With this feature in place, I could choose to update other packages while keeping `titiler-pgstac<0.7`. Without this feature I would need to stick with whichever dependency versions that got installed alongside `titiler-pgstac<0.7` in an `eoapi-cdk` release.

The only downside that I can see about letting the user pick specific versions is the `handler` problem. If there are version-specific tweaks that need to be made to the handlers (like we probably need to do for `pgstac>=0.8`) then `eoapi-cdk` would need to have a way to declare a range of dependency versions that are compatible with a given release.

**Comment by vincentsarago** [comment]
We so many runtime version I'm worry that it will be confusing and that most user will not get that handlers are dependent to a specific version! I feel that having a default in eoapi-cdk the most `safe` way, as long as we make customization (letting user pass their own dockerfile and handler) 

imagine that you don't have eoapi-cdk pinned to a specific version but you have an environment settings that overwrite the default titiler-pgstac, but in a new release we `break` the handler, you won't get any notification about why your application doesn't work anymore! It will be messy to debug!

**Comment by hrodmn** [comment]
> This seems a lot to add here, It adds a burden to the maintainer of this library for something which (IMO) is better and safer to be handled as user level (via customization)

@vincentsarago @emileten it's totally understandable that you would not want to keep track of compatibility with all possible versions of packages like `titiler-pgstac` in this package. If the custom Dockerfile/handler path is the best way forward I will do that!

**Comment by emileten** [comment]
> This seems a lot to add here

I would agree with this statement from Vincent. It would be great but it's a bit risky because we don't have much time to maintain this library.
